train_model.py

```diff
@@ -0,0 +1,420 @@
+            train_total_loss += loss * len(inputs)
+            train_total_samples += len(inputs)
+
+            # Backpropagate loss
+            loss.backward()
+
+            # Update parameters
+            optimizer.step()
+
+            # Run validation 3 times per epoch around 33%, 66%, and 100%
+            if (3*(batch_idx + 1)/len(train_loader) % 1) <= 2/len(train_loader):
+                # Print average train loss
+                avg_train_loss = train_total_loss / train_total_samples
+                print("Average Train Loss Since Last Validation Run: " + \
+                      f"{avg_train_loss}")
+                train_total_loss = 0
+                train_total_samples = 0
+
+                print("Taking a break to run validation....")
+                model.eval()
+                val_total_loss = 0
+                val_total_samples = 0
+                with torch.inference_mode():
+                    for val_inputs, val_targets in tqdm(valid_loader,
+                                                        desc="Validate"):
+                        # Put validation inputs and targets on device
+                        val_inputs = val_inputs.to(device, non_blocking=True)
+                        val_targets = val_targets.to(device, non_blocking=True)
+
+                        # Get validation predictions
+                        val_predictions = model(val_inputs)
+
+                        # Reshape the model predictions for Cross Entropy
+                        val_predictions = val_predictions.transpose(-2, -1)
+
+                        # Calculate validation loss
+                        val_loss = loss_fn(val_predictions, val_targets)
+                        val_total_loss += val_loss.item() * len(val_inputs)
+                        val_total_samples += len(val_inputs)
+
+                    # Print average validation loss
+                    avg_val_loss = val_total_loss / val_total_samples
+                    print(f"\nAverage Validation Loss: {avg_val_loss}")
+
+    # Test the model
+    print("\nDone training! Now testing model...")
+    model.eval()
+    total_loss = 0
+    total_samples = 0
+    with torch.inference_mode():
+        for inputs, targets in tqdm(test_loader, desc="Test"):
+            # Put inputs and targets on device
+            inputs = inputs.to(device, non_blocking=True)
+            targets = targets.to(device, non_blocking=True)
+
+            # Get model predictions
+            predictions = model(inputs)
+
+            # Reshape the model predictions for Cross Entropy
+            predictions = predictions.transpose(-2, -1)
+
+            # Calculate loss
+            loss = loss_fn(predictions, targets)
+            total_loss += loss.item() * len(inputs)
+            total_samples += len(inputs)
+
+    # Print average testing loss
+    avg_loss = total_loss / total_samples
+    print(f"Average Test Loss: {avg_loss}")
+
+    # Generate text from the model
```

Comment:
TODO: Remove generating text section, replace with saving model. (Perhaps save generating text section for sanity check at end of training, but this is really the responsibility of inference in utils.py, not train_model.py).